### PR TITLE
VMware: Refactor vmware_host module

### LINKED ISSUE #20819: Error on vmware_host : 'vim.fault.DuplicateName' object has no attribute 'thumbprint'
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
modules/cloud/vmware/vmware_host

##### ANSIBLE VERSION

```
2.2.1.0
```

##### OS / ENVIRONMENT
Ansiblehost: Ubuntu 16.04
ESX/VSphere: 6.5

##### SUMMARY
Trying to add multiple ESX Host to a VSphere Cluster

##### STEPS TO REPRODUCE

```yaml
- name: "Ensure ESX Host is in vSphere Center and in the vSphere Cluster VMC1"
  hosts: vsphere-vmc1-esx
  connection: local
  vars:
    cluster: "{{vsphere_cluster_default}}"
    datacenter: "{{vsphere_datacenter_default}}"
    vcenterserver: "{{vsphere_default_vcenter}}"
    vsphereuser: "{{vsphere_vmc1_user}}"
    vspherepw: "{{vsphere_vmc1_password}}"
  tasks:
    - name: "debug"
      debug:
        var: vsphere_vmc1_user
    - name: "Add to vCenter"
      vmware_host:
        cluster_name: "{{cluster}}"
        datacenter_name: "{{datacenter}}"
        esxi_hostname: "{{ansible_host}}"
        esxi_username: "{{esxi_username}}"
        esxi_password: "{{esxi_password}}"
        hostname: "{{vcenterserver}}"
        username: "{{vsphereuser}}"
        password: "{{vspherepw}}"
        validate_certs: False
        state: present

```

##### ACTUAL RESULTS
Looks like something is missing in the Module to handle the thumbprint attribute

Looks like same Issue like https://github.com/ansible/ansible-modules-extras/issues/1484
but not resolved.

```
fatal: [fj-p-vmnode1]: FAILED! => {

    "changed": false,
    "failed": true,
    "invocation": {
        "module_args": {
            "cluster_name": "vmc1",
            "datacenter_name": "FJ",
            "esxi_hostname": "172.16.125.1",
            "esxi_password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "esxi_username": "root",
            "hostname": "172.16.125.200",
            "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "state": "present",
            "username": "ansible@vsphere.local",
            "validate_certs": false
        },
        "module_name": "vmware_host"
    },
    "msg": "'vim.fault.DuplicateName' object has no attribute 'thumbprint'"
}
```

### PR DESCRIPTION
**Pull request by Akasurde**
##### SUMMARY
This fix refactor vmware_host module to use PyVmomi.
Also, handle SSLVerifyFault exception to get hostsystem's certificate
thumbprint.

Fixes: #20819

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
lib/ansible/modules/cloud/vmware/vmware_host.py

##### ANSIBLE VERSION

```
2.5devel
```

### CODE REVIEW

**Comment by Akasurde** [comment]
cc @tchernomax

**Comment by tchernomax** [comment]
@Akasurde I now understand why and what exactly you asked me:

> can you paste me TaskError raised while connecting host in vcenter ? I want to see if it contains thumbprint or not this is related  to vmware_host

However, I just checked and I don't have any preproduction vcenter where I can test that… I will ask my colleagues tomorrow to know if they have any spare/test machine.

The PR #30582 was test on "production" machines previously freed of all their VM (so they were not "production" machines during my tests) during special operations ; we where upgrading all our ESXi.
Those operations are now finished… and I can't VMotion all the VM of an ESXi in production just to test the content of TaskError object.

But as I said, I will ask my colleagues tomorrow.

**Comment by Akasurde** [comment]
cc @pdellaert

**Comment by tchernomax** [comment]
@Akasurde I tried but couln't get any vim.fault.SSLVerifyFault error.

File: lib/ansible/modules/cloud/vmware/vmware_host.py [hunk 1]
```diff
@@ -107,99 +114,107 @@
...
+                # In order to get the sslThumbprint we first connect
+                # get the vim.fault.SSLVerifyFault then grab the sslThumbprint
+                # from that object.
+                #
+                # args is a tuple, selecting the first tuple
+                ssl_verify_fault = add_task_error.args[0]
+                host_connect_spec.sslThumbprint = ssl_verify_fault.thumbprint
 
         task = self.cluster.AddHost_Task(host_connect_spec, as_connected, resource_pool, esxi_license)
         success, result = wait_for_task(task)
```
Comment (tchernomax) [inline lib/ansible/modules/cloud/vmware/vmware_host.py]: I would have write:
```
-        try:
-            task = self.cluster.AddHost_Task(host_connect_spec, as_connected, resource_pool, esxi_license)
-            success, result = wait_for_task(task)
-            return success, result
-        except TaskError as add_task_error:
-            # This is almost certain to fail the first time.
-            # In order to get the sslThumbprint we first connect
-            # get the vim.fault.SSLVerifyFault then grab the sslThumbprint
-            # from that object.
-            #
-            # args is a tuple, selecting the first tuple
-            ssl_verify_fault = add_task_error.args[0]
-            host_connect_spec.sslThumbprint = ssl_verify_fault.thumbprint
-
-        task = self.cluster.AddHost_Task(host_connect_spec, as_connected, resource_pool, esxi_license)
-        success, result = wait_for_task(task)
-        return success, result
+        for _ in range(0, 2):
+            try:
+                task = self.cluster.AddHost_Task(host_connect_spec, as_connected, resource_pool, esxi_license)
+                success, result = wait_for_task(task)
+                return success, result
+
+            except TaskError as task_error_exception:
+
+                task_error = task_error_exception.args[0]
+                if self.esxi_ssl_thumbprint == '' and isinstance(task_error, vim.fault.SSLVerifyFault):
+                    # User has not specified SSL Thumbprint for ESXi host,
+                    # try to grab it using SSLVerifyFault exception
+                    host_connect_spec.sslThumbprint = task_error.thumbprint
+                else:
+                    self.module.fail_json(msg="Failed to add host %s to vCenter: %s" % (self.esxi_hostname, task_error.msg))
+
+        self.module.fail_json(msg="Failed to add host %s to vCenter: %s" % (self.esxi_hostname, task_error.msg))
```

It seems vim.fault.SSLVerifyFault and other vim.fault… are always packed in a TaskError.

File: lib/ansible/modules/cloud/vmware/vmware_host.py [hunk 2]
```diff
@@ -216,16 +231,19 @@ def reconnect_host_to_vcenter(self):
...
+                    return success, result
+                except TaskError as e:
+                    self.module.fail_json(msg="Failed to reconnect host %s to vCenter due to %s" % (self.esxi_hostname, e))
+                except vim.fault.SSLVerifyFault as add_task_error:
+                    # See add_host_to_vcenter
+                    ssl_verify_fault = add_task_error.args[0]
+                    reconnecthost_args['cnxSpec'].sslThumbprint = ssl_verify_fault.thumbprint
 
         task = self.host.ReconnectHost_Task(**reconnecthost_args)
         success, result = wait_for_task(task)
```
Comment (tchernomax) [inline lib/ansible/modules/cloud/vmware/vmware_host.py]: Same here:
```
+            for _ in range(0, 2):
+                try:
+                    task = self.host.ReconnectHost_Task(**reconnecthost_args)
+                    success, result = wait_for_task(task)
+                    return success, result
+
+                except TaskError as task_error_exception:
+
+                    task_error = task_error_exception.args[0]
+                    if self.esxi_ssl_thumbprint == '' and isinstance(task_error, vim.fault.SSLVerifyFault):
+                        # User has not specified SSL Thumbprint for ESXi host,
+                        # try to grab it using SSLVerifyFault exception
+                        reconnecthost_args['cnxSpec'].sslThumbprint = task_error.thumbprint
+                    else:
+                        self.module.fail_json(msg="Failed to reconnect host %s to vCenter: %s" % (self.esxi_hostname, task_error.msg))

+               self.module.fail_json(msg="Failed to reconnect host %s to vCenter: %s" % (self.esxi_hostname, task_error.msg))
+
+        else:
             try:
                 task = self.host.ReconnectHost_Task(**reconnecthost_args)
                 success, result = wait_for_task(task)
                 return success, result
-            except TaskError as add_task_error:
-                # See add_host_to_vcenter
-                ssl_verify_fault = add_task_error.args[0]
-                reconnecthost_args['cnxSpec'].sslThumbprint = ssl_verify_fault.thumbprint
-
-        task = self.host.ReconnectHost_Task(**reconnecthost_args)
-        success, result = wait_for_task(task)
-        return success, result
+            except TaskError as task_error_exception:
+                task_error = task_error_exception.args[0]
+                self.module.fail_json(msg="Failed to reconnect host %s to vCenter due to %s" % (self.esxi_hostname, task_error.msg))
```

**Review by tchernomax — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/cloud/vmware/vmware_host.py (hunk 1 above)
Comment (tchernomax) [reply lib/ansible/modules/cloud/vmware/vmware_host.py]: @Akasurde what do you think ?

**Comment by Akasurde** [comment]
@tchernomax Thanks, I have integrated your suggestions.

**Review by pdellaert — APPROVED** [review]
LGTM
shipit

**Review by tchernomax — APPROVED** [review]
LGTM

**Comment by Akasurde** [comment]
@tchernomax @pdellaert Thanks for review and merge.

**Pull request merged**